[로또 - 웹, DB] 이바 미션 제출합니다.

seonghun127 opened the pull request:
많이 부족하지만 MVC 패턴을 적용하려고 노력했습니다.
전체적으로 웹 UI 보단 서버 구현에 신경을 썼습니다. 
Controller, Service, Dao와 같이 3 layer를 나눠보는 연습을 많이 할 수 있었습니다.
부족한 부분은 가차없이 지적해주시면 감사하겠습니다.

seonghun127 committed:
docs : 기능 구현 목록 갱신

seonghun127 committed:
docs : 기능 구현 목록 갱신

seonghun127 committed:
refactor : LottoFactory 클래스의 createLottos 메소드의 반환 타입 변경

1. 반환 타입을 Lottos에서 List<Lotto>로 변경한다
- 반환 타입을 Lottos로 하게됐을 경우 Lottos 타입으로 한정정이게 된다.
이를 List<> 타입으로 변경함으로써 사용될 수 있는 경우를 늘려 유연하게 만든다.

seonghun127 committed:
feat : 자 / 수동으로 생성된 로또 번호 리스트를 해당 회차 정보와 저장하는 DAO 구현

seonghun127 committed:
refactor : LottoFactory 클래스의 createLottos 메소드의 반환 타입 변경으로 인한 테스트 코드 수정

seonghun127 committed:
refactor : DAO 클래스에 try-with-resources 로직 적용

seonghun127 committed:
refactor : 데이터베이스 연결 클래스 따로 분리

seonghun127 committed:
feat : Round 를 관리하는 테이블 추가 및 DAO 구현

1. round 테이블은PK로 round를 가지며 round별 구입 금액(amount) 칼럼을 둬서 데이터를 저장한다
2. lotto / winning_lotto 테이블은 round 테이블의 round 값을 가지고 해당 회차 정보임을 저장하고 확인한다

seonghun127 committed:
feat : winning_lotto 테이블 추가 및 DAO 구현

1. winning_lotto 테이블은 당첨 번호, 보너스 번호, 라운드 정보를 저장한다
2. 당첨 번호, 보너스 번호를 찾는 메서드를 따로 구분한다
- 뷰에서 당첨 번호와 보너스 번호를 따로 전달 받고 이 둘의 검색 결과를 가지고 WinningLotto 객체를 만들기 위해

seonghun127 committed:
feat : StringUtil commons lang 라이브러리 추가

seonghun127 committed:
docs : 기능 구현 목록 갱신

seonghun127 committed:
feat : 내림 차순으로 round를 모두 출력하는 기능 추가

seonghun127 committed:
refactor : Controller로 로직을 분리하고 코드 간소화

seonghun127 committed:
feat : 입력 받은 로또 번호를 바탕으로 로또 리스트 생성

1. Controller는 view에서 입력 받은 인자를 바로 LottoService를 호출하면서 넘겨준다
2. LottoService를 통해 Controller의 복잡한 로직을 분리한다
3. LottosDTO를 가지고 해당 기능에 따른 view 에서 필요한 데이터를 담은 클래스를 생성하고 반환한다

seonghun127 committed:
feat : round를 생성하고 굼액을 저장하는 기능 추가

1. RoundController를 가지고 view에서 넘어온 인자를 service에게 넘겨준다
2. RoundService 에서 넘겨받은 인자를 바탕으로 구매금액 저장, 전체 라운드 내림차순 검색,
라운드 최대값 검색 기능을 구현한다

seonghun127 committed:
feat : 당첨 번호를 저장, 당첨 결과를 계산하고 결과를 반환하는 기능 추가

1. Controller는 view에서 넘어온 인자를 바로 service에게 넘겨준다
2. WinningLottoService는 넘겨받은 인자를 가지고 당첨 번호 저장, 당첨 결과 계산하는 기능을 구현한다
3. 저장된 당첨 번호를 나타내는 WinningLottoDTO 객체를 생성하여 반환한다
4. 계산된 당첨 결과를 나타내는 WinningResultDTO 객체를 생성하여 반환한다

seonghun127 committed:
feat : 수동 로또, 당첨 번호 로또를 생성시 view에서 입력받는 문자열을 list 바꾸는 중복 로직을 따로 분리

seonghun127 committed:
feat : 로또 구매를 입력받는 view 구현

seonghun127 committed:
feat : 로또 구매 개수를 출력하고 수동 구매 개수, 번호를 입력받는 view 구현

seonghun127 committed:
feat : 로또 번호 리스트를 출력하고 당첨 번호, 보너스 번호를 입력받는 view 구현

seonghun127 committed:
당첨 번호, 보너스 번호를 출력하고 회차를 선택하여 결과를 조회할 수 있는 view 구현

seonghun127 committed:
feat : 선택된 회차의 결과를 출력하는 view 구현

seonghun127 committed:
refactor : dao 메소드 명과 service 메소드 명을 일치

seonghun127 committed:
refactor : String 처리 중복 메서드를 StringUtil 클래스로 추출

gracefulBrown commented on src/main/java/lotto/Controller/RoundController.java:
```diff
@@ -0,0 +1,31 @@
+package lotto.Controller;
+
+import lotto.dao.RoundDAO;
```
Optimize imports 기능을 사용해보세요 :)

seonghun127 replied:
반영했습니다!

gracefulBrown commented on src/main/java/lotto/WebUILottoApplication.java:
```diff
@@ -1,22 +1,29 @@
-            return render(model, "index.html");
-        });
+
+        get("/", RoundController.MAX_ROUND);
```
아래의 Controller들의 function들은 행위를 의미하는 네이밍인것과 달리 MAX_ROUND만 명사같네요 ㅎㅎ;;

seonghun127 replied:
`GET_MAX_ROUND` 라는 이름으로 변경했습니다.
저는 Controller에 있는 필드들이 상수라고 생각하는데 보니까 동사로 시작하는 네이밍을 했더군요...
이들도 하나의 객체이기때문에 동사로 이름을 짓는것은 convention에 어긋나는 것이라고 생각합니다.
근데 이름으로 이들의 역할을 명확히하려면 지금처럼 동사가 들어가는게 좋다고 생각도 듭니다.
이 부분에 대해서는 어떻게 생각하시나요?

gracefulBrown commented on src/main/java/lotto/service/LottoService.java:
```diff
@@ -0,0 +1,39 @@
+    private final RoundDAO roundDAO = new RoundDAO();
+    private final LottoDAO lottoDAO = new LottoDAO();
+
+    public LottosDTO.Create createLottos(String manualLottoNumbers) throws SQLException {
```
WebUiController와 ConsoleController 모두 LottoService를 사용해야 하는것 같은데 manualLottos 에 대한 전처리는 컨트롤러 레이어에서 하면 어떨까요?

seonghun127 replied:
Console 방식과 Web 방식이 view를 어떤 형식(콘솔이냐, 웹 브라우져냐)으로 출력하나의 차이만 있고 내부 로직이 돌아가는 방식은 같다고 생각해서 피드백 주신 방식대로 수정했습니다.
구현한 Service 클래스도 LottoApplication (Console 방식)에서도 잘 돌아가야하기에 service의 `createLotto()` 와 `createWinningLotto()` 메서드가 모두 `List<String>` 타입으로 인자를 받는 것이 맞다고 생각이 들었습니다.
근데 여기서 궁금한 점은 Controller 단에서 view 에서 넘어온 인자를 (형변환과 같은)가공을 해도 상관이 없는 것인가요? 만약 가공하는 과정에서 예외가 발생하면 그 부분을 처리하는 로직도 추가해야할텐데 그러면 Controller 가 하는 역할이 불어나는것이 아닌가라는 생각이 들었습니다.

gracefulBrown commented on src/main/java/lotto/service/LottoService.java:
```diff
@@ -0,0 +1,39 @@
+    private final LottoDAO lottoDAO = new LottoDAO();
+
+    public LottosDTO.Create createLottos(String manualLottoNumbers) throws SQLException {
+        logger.info("create lotto!");
```
log를 남기는 목적이 디버깅의 목적이라면
디버깅 툴을 이용하는 것을 추천드립니다 :)

seonghun127 replied:
반영했습니다!

gracefulBrown commented on src/main/java/lotto/LottoApplication.java:
```diff
@@ -14,9 +14,9 @@
                 setManualLottoNumbers(setManualPurchaseCount(purchaseAmount / LOTTO_PRICE)),
-                purchaseAmount / LOTTO_PRICE);
+                purchaseAmount / LOTTO_PRICE));
         OutputView.outputLottos(lottos);
```
lottos의 toString에 출력 로직을 정의할 경우 ConsoleUI에 종속될 수 있습니다 ㅠㅠ

seonghun127 replied:
`Lottos` 클래스에 `getLottos()` 메서드를 추가하고 이를 view에 전달하는 방식으로 지적해주신 부분을 해결했습니다.
피드백을 반영하면서 궁금한 사항이 생겼습니다. 기존에 `Lottos`  클래스의 `toString()` 메서드를 재정의하고 이를 view에서 사용하는 방식이었습니다.
```
StringBuilder sb = new StringBuilder();
for (Lotto lotto : lottos) {
     sb.append(lotto);
     sb.append("\n");
}
return sb.toString();
```
근데 위 코드에서 `sb.append(lotto)` 부분도 `Lotto` 클래스에서 `toString()` 메서드를 재정의하고 사용하는 방식인데, 그럼 그 부분도 view에 종속적인 코드가 되는 건가요? `Lotto` 클래스의 `toString()` 메서드는 로또 번호가 [1, 2, 3, 4, 5, 6] 이런 식으로 출력하기 위해 구현한 상태입니다.
`toString()` 메서드를 재정의하는 이유는 해당 클래스의 정보를 유용하게 출력하기 위함이라고 알고있습니다. 그래서 `Lotto` 클래스에서 로또 6자리 번호가 모두 출력될 수 있도록 재정의를 한 것인데 그 부분을 콘솔 view 에서 출력할 때 사용하는 것도 view에 종속적이게 되는 것인지가 궁금하네요.

gracefulBrown commented on src/main/java/lotto/service/LottoService.java:
```diff
@@ -0,0 +1,39 @@
+        return new LottosDTO.Create(lottos);
+    }
+
+    public LottosDTO.Create findLottosByRound(int round) throws SQLException {
```
SQLException처리는 어디서 하나요?
Exception을 발생 시 처리 로직도 포함되면 좋을 것 같아요 :)

seonghun127 replied:
반영했습니다!

gracefulBrown reviewed (changes requested):
안녕하세요. 리뷰어 브라운 입니다.
웹UI와 DB연동을 하면서 전체적으로 코드가 많이 추가되었네요!
전체적인 구조나 레이어 분리 등 고민을 많이 하신것을 보입니다 👍 
몇가지 피드백포함하여 코멘트 남겼으니 확인해보세요~
궁금한 점 있으면 언제든 DM 주세요 :)

seonghun127 committed:
refactor : import 최적화

seonghun127 committed:
refactor : Lottos 클래스에 getLottos 메서드 추가

1. 기존의 toString 메서드로 view에서 출력하던 로직을 변경한다
- view에 종속적인 문제를 해결하기 위해
2. view에 get 메서드로 Lottos 클래스가 가지고 있는 인스턴스 필드를 전달한다
- Collections.unmodifiableList 메서드로 read-only의 필드를 반환한다

seonghun127 committed:
refactor : import 최적화

seonghun127 committed:
refactor : RoundController의 상수명 변경

1. MAX_ROUND 상수를 행위를 의미하는 네이밍(GET_MAX_ROUND)으로 변경한다

seonghun127 committed:
refactor : log 삭제

seonghun127 committed:
refactor : SQLException을 DAO에서 처리

1. SQLException 처리를 해주는 곳이 없어 직접 쿼리를 실행시키는 DAO에서 try-with-resources에 catch를 추가하여 처리한다

seonghun127 committed:
refactor : LottoFactory 클래스의 createLottoManually 메서드를 사용하는 Service 클래스의 메서드에서 convertToList 메서드 제거

1. String으로 인자를 받던 방식에서 List<String> 인자를 받는 방식으로 변경한다
- List로 변환하는 로직을 Controller에서 수행하는 방식으로 변경하기 때문
- Web과 Console 어플리케이션의 방식 모두 충족하기 위해
- service 객체를 생성하고 메서드를 호출하는 레이어에서 String 인자를 List로 변경하고 넘겨준다 (Console & Web)

seonghun127 committed:
refactor : LottoApplication (ConsoleController)의 main 메서드에서 Service 클래스 적용

1. 웹과 같은 작동을 하는 Console Application(LottoApplication)에 service 레이어를 사용한다
2. service 레이어단 객체를 호출하고 메서드를 호출하면서 Console Application에서도 DB에 접근한다

seonghun127 commented:
피드백을 전부 반영하려고 노력했습니다.
console application에서도 db에 저장하는 방식으로 변경했습니다.
추가로 손 볼 필요가 있는 부분은 많이많이 지적해주세요! 감사합니다 👍

gracefulBrown commented on src/main/java/lotto/domain/LottoFactory.java:
```diff
@@ -17,7 +17,7 @@ public static Lotto createLottoManually(List<String> manualNumbers) {
     }
 
-    public static Lottos createLottos(List<String> manualLottoNumbers, int countOfPurchase) {
+    public static List<Lotto> createLottos(List<String> manualLottoNumbers, int countOfPurchase) {
```
현재 Lottos는 WinningLotto관련해서만 사용하고 있나요?
createLottos의 리턴으로 Lottos를 응답하면 어떨지 고민해 보면 좋을 것 같아요 :)

seonghun127 replied:
적용했습니다!

gracefulBrown commented on src/main/java/lotto/dao/LottoDAO.java:
```diff
@@ -0,0 +1,64 @@
+
+        try (Connection connection = DatabaseConnection.getConnection();
+             PreparedStatement statement = connection.prepareStatement(sql)) {
+            int nextRound = new RoundDAO().findMaxRound();
```
Dao에서 다른 DAO를 호출 한다면 Dao간 순환참조 현상이 발생 될 가능성이 높습니다.
nextRound값은 파라미터로 넘겨받는 걸 추천합니다 :)

seonghun127 replied:
적용했습니다!

gracefulBrown commented on src/main/java/lotto/dao/WinningLottoDAO.java:
```diff
@@ -0,0 +1,81 @@
+
+        try (Connection connection = DatabaseConnection.getConnection();
+             PreparedStatement statement = connection.prepareStatement(sql)) {
+            int nextRound = new RoundDAO().findMaxRound();
```
Dao, Service의 경우 상태가 없어서 매번 새로 생성해줄 필요가 없습니다.
싱글톤 패턴으로 구현하는거에 대해서 고민해보세요 ^^

seonghun127 replied:
적용했습니다!

gracefulBrown reviewed (approved):
피드백 반영 잘 해주셨습니다 👍 
몇가지 피드백 남겼으니 확인해주세요 ^^
추가적인 부분은 오프라인으로 전달하겠습니다 :)

seonghun127 committed:
fix: DAO 클래스에 모두 싱클톤 패턴 적용

seonghun127 committed:
fix: Service 클래스에 싱글톤 패턴 적용

seonghun127 committed:
fix: DAO에서 다른 DAO를 호출하는 로직을 제거

1. 다른 DAO를 호출해서 사용했던 값은 파리미터로 넘긴다
- 순환 참조의 위험을 막기위해

seonghun127 committed:
Service의 싱글톤 패턴 적용으로 Service 인스턴스 호출 로직 변경

seonghun127 committed:
fix: LottoFactory 클래스의 createLottos() 메서드 반환 타입 변경

1. List<Lotto>에서 Lottos로 변경한다
- Lottos 객체를 생성해서 반환하는 것이 Factory 라는 이름에 일맥상통하기 때문

seonghun127 committed:
refactor: 코드 수정에 따른 추가 변동사항 적용

seonghun127 commented:
싱글톤 패턴 적용방법을 찾아보면서 이전에 적용했던 방식과 다른 방식으로 적용해봤습니다. 
Holder 이름의 inner class로 두는 방식을 공부할 수 있었습니다.

순환 참조에 대해 공부할 수 있었습니다. DAO, Service layer에서 순환 참조가 일어나지 않게 프로그램을 짜야한다는 것을 알게되었습니다.

gracefulBrown commented on src/main/java/lotto/service/LottoService.java:
```diff
@@ -0,0 +1,35 @@
+
+import java.util.List;
+
+public class LottoService {
```
Service 로직에 대한 테스트코드가 없네요ㅠㅠ

gracefulBrown commented on src/main/java/lotto/dao/DatabaseConnection.java:
```diff
@@ -0,0 +1,34 @@
+package lotto.dao;
```
dao보다는 util쪽으로 옮기는게 좋을 것 같아요 :)

gracefulBrown commented on src/main/java/lotto/dao/LottoDAO.java:
```diff
@@ -0,0 +1,73 @@
+        String sql = "SELECT lotto FROM lotto WHERE round = ?";
+        List<Lotto> lottos = new ArrayList<>();
+
+        try (Connection connection = DatabaseConnection.getConnection();
```
DAO로직에서 코드 중복이 많이 발생하고 있네요 ^^
다음에 시도할 때 중복제거도 같이 고려해보면 좋을 것 같아요 :)

gracefulBrown reviewed (approved):
피드백 반영하시느라 고생하셨습니다 👍 
로직에 비해 테스트 코드가 적은것 같아서 아쉽지만 전체적으로 구조가 잘 짜여진것같아요 :)
몇가지 도움이 될 수 있는 피드백 남기니 다음 미션에서 고려해보면 좋을 것 같아요!
수고하셨습니다.
